### PR TITLE
Basic multiwallet support

### PR DESCRIPTION
**Pull request by luke-jr**
This allows running with multiple -wallet options to load more than one wallet.

~~GUI has independent comboboxen for the main GUI window and debug console to select which wallet to view/use. The comboboxes are not visible in the main GUI unless multiple wallets are loaded (the debug window's combobox is useful even for a single wallet, since it allows selecting "(none)" to block wallet access).~~

~~RPC can access only one wallet per user, but rpcauth is extended to accept a 4th field which controls which wallet, if any, the user has access to. I chose to do it this way because pre-multiwallet nodes will gracefully ignore such rpcauth rather than expose their wallet. The field can also be a single hyphen to block wallet access.~~

(RPC and GUI changes are moved to my `multiwallet_rpc` and `multiwallet_gui` branches)

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Impressive changeset!
General Concept ACK.

Instead of the extending the RPC auth, I could imaging using different RPC endpoints would also work.
* Accessing RPC at `/` would result in the default wallet.
* Accessing RPC at `/<walletid>` would result in using the wallet with the id `<walletid>`.

bitcoin-cli could just support a new argument `-walletid`.
Something like `bitcoin-cli -walletid=something getbalance` would result in the same RPC call just against the endpoint `/something`.
Maybe this would result in a simple RPC access to the multiple wallets.

Will review soon.

**Comment by luke-jr** [comment]
> Instead of the extending the RPC auth, I could imaging using different RPC endpoints would also work.

Indeed, although my main purpose in doing this was to isolate JoinMarket, and it's less isolated if it has auth to my real hotwallet. :)

(Also, endpoints seemed like they'd require more code.)

> Accessing RPC at `/<walletid>` would result in using the wallet with the id `<walletid>`.

This isn't very future-proof. If we go this route, I'd suggest `/?wallet=<walletid>`

**Comment by btcdrak** [comment]
Nice, Concept ACK. Will test.

File: src/init.cpp [hunk 1]
```diff
@@ -196,8 +196,9 @@ void Shutdown()
     StopRPC();
     StopHTTPServer();
 #ifdef ENABLE_WALLET
-    if (pwalletMain)
-        pwalletMain->Flush(false);
+    for (CWallet_ptr pwallet : vpwallets) {
```
Comment (MarcoFalke) [inline src/init.cpp]: I like the approach by @jonasschnelli better. Let's keep most of the multiwallet logic in `wallet.cpp`.

**Comment by luke-jr** [comment]
Rebased, but please prioritise #8775 review (which this is based on) first. Once that's merged, it may or may not make sense to split the Qt changes out of this.

**Comment by gmaxwell** [comment]
Concept ACK.

**Comment by luke-jr** [comment]
Moved RPC/GUI changes out of this, and rebased.

File: src/init.cpp (hunk 1 above)
Comment (luke-jr) [inline src/init.cpp]: It's not in `wallet.cpp` now. Moving it there seems like a good idea, but independent of this PR...

**Comment by luke-jr** [comment]
Rebased and now the next-step in multiwallet integration.

**Comment by jonasschnelli** [comment]
Re-Concept ACK.
Binaries to play with: https://bitcoin.jonasschnelli.ch/build/27

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -819,21 +819,23 @@ void ThreadFlushWalletDB()
...
-                        nLastFlushed = CWalletDB::GetUpdateCounter();
-                        int64_t nStart = GetTimeMillis();
-
-                        // Flush wallet file so it's self contained
-                        bitdb.CloseDb(strFile);
-                        bitdb.CheckpointLSN(strFile);
-
-                        bitdb.mapFileUseCount.erase(_mi++);
-                        LogPrint("db", "Flushed %s %dms\n", strFile, GetTimeMillis() - nStart);
+                    for (CWallet_ptr pwallet : vpwallets) {
```
Comment (ryanofsky) [inline src/wallet/walletdb.cpp]: This thread is created from the non-static method `CWallet::postInitProcess`, so it seems like it makes more sense for it to continue flushing a single wallet, rather than going through the global wallets list. Instead of looping and accessing the global `vpwallet` variable, `ThreadFlushWalletDB` could just take a new `CWallet* pwallet` argument and be invoked with:

```
threadGroup.create_thread([this](){ThreadFlushWalletDB(this);});
```

from `CWallet::postInitProcess`.

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -819,21 +819,23 @@ void ThreadFlushWalletDB()
...
+
+                            // Flush wallet file so it's self contained
+                            bitdb.CloseDb(strFile);
+                            bitdb.CheckpointLSN(strFile);
+
+                            bitdb.mapFileUseCount.erase(_mi++);
+                            LogPrint("db", "Flushed %s %dms\n", strFile, GetTimeMillis() - nStart);
+                        }
                     }
+                    nLastFlushed = CWalletDB::GetUpdateCounter();
```
Comment (ryanofsky) [inline src/wallet/walletdb.cpp]: How come the `nLastFlushed` assignment is moving ouside of the `_mi != bitdb.mapFileUseCount.end()` condition? It seems unrelated to this change. Maybe a code comment explaining the assignment would be good here.

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -31,7 +31,8 @@
 #include <boost/shared_ptr.hpp>
 #include <boost/thread.hpp>
 
-extern CWallet* pwalletMain;
+typedef CWallet* CWallet_ptr;
```
Comment (luke-jr) [inline src/wallet/wallet.h]: Indeed, a shared_ptr or similar would be needed for when/if we add the ability to close wallets at runtime.

File: src/wallet/walletdb.cpp (hunk 2 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Having a dedicated thread per wallet is just needlessly inefficient.

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Because the condition is per-wallet. Maybe nLastFlushed ought to be as well...?

**Comment by ryanofsky** [comment]
(needs another rebase)

**Comment by laanwj** [comment]
Added 0.15.0 milestone.

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Sanitised this better in the rebase

**Comment by luke-jr** [comment]
Significantly rebased. Also fixed minor race conditions surrounding wallet flush (it was bumping the "have update to flush" counter before doing the actual update).

**Comment by laanwj** [comment]
It would be preferable to increment the update counter after committing a database transaction, not after every call to CWalletDB.
So this would mean the destructor of `CWalletDB` and `TxnCommit()` . Which should be overridden in CWalletDB, or alternatively use composition not inheritance for CDB as I've done in the last commit in #9951.
This will simplify these changes (no need to do WriteIC/EraseIC everywhere), as well as makes sure that only changes that actually hit the disk are counted.

File: src/util.cpp [hunk 5]
```diff
@@ -418,6 +418,7 @@ bool SoftSetArg(const std::string& strArg, const std::string& strValue)
     if (mapArgs.count(strArg))
         return false;
     mapArgs[strArg] = strValue;
+    _mapMultiArgs[strArg].push_back(strValue);
```
Comment (laanwj) [inline src/util.cpp]: I think it is confusing for a "Set" function to add something to a list. If anything it should replace it.
What do you need this for, specifically?
In the past I've defined [SoftSetMultiArg](https://github.com/laanwj/bitcoin/commit/1024183eff629b758745aa7ebf9329d987179644) to be able to define a whole list argument at once.

File: src/wallet/test/wallet_test_fixture.cpp [hunk 6]
```diff
@@ -8,6 +8,8 @@
 #include "wallet/db.h"
 #include "wallet/wallet.h"
 
+CWallet *pwalletMain;
```
Comment (laanwj) [inline src/wallet/test/wallet_test_fixture.cpp]: Why define pwalletMain here?

**Comment by laanwj** [comment]
I think the overall approach here is good. This is another step towards multi-wallet support. As it's not anywhere on the API yet, I guess it's too early to ask for a multiwallet test?

File: src/util.cpp (hunk 5 above)
Comment (luke-jr) [inline src/util.cpp]: It can't get here unless the list is currently empty. It's used to initialise `-wallet`, which is accessed as a list only now.

File: src/wallet/test/wallet_test_fixture.cpp (hunk 6 above)
Comment (luke-jr) [inline src/wallet/test/wallet_test_fixture.cpp]: There's no need to modify the tests to use vpwallets in most cases?

**Comment by luke-jr** [comment]
> It would be preferable to increment the update counter after committing a database transaction, not after every call to a CWalletDB method.

IMO too much for this PR. Just trying to multiwallet here, not refactor stuff that doesn't need it. :)

File: src/wallet/test/wallet_test_fixture.cpp (hunk 6 above)
Comment (laanwj) [inline src/wallet/test/wallet_test_fixture.cpp]: Depends on the reason why they are setting pwalletMain. Generally that is to communicate with stuff in wallet.cpp/walletdb.cpp I'd say.

**Comment by laanwj** [comment]
> IMO too much for this PR. Just trying to multiwallet here, not refactor stuff that doesn't need it. :)

Maybe my recommendation goes to far but I just think your current approach there is messy. The mutator functions call `WriteIC`/`EraseIC` which call `IncrementUpdateCounter` which calls `IncrementUpdateCounter(strFile)` which indexes into a global map indexed by name. This makes the wallet code even more complex and hard to understand as it introduces another place where per-wallet(-database) information is stored.

We need to decide whether CWalletDBStateInfo belongs with the database wrapper, or with the wallet.

- If it is part of the database statistics side of things, the place to put it would be CDBEnv, with the other maps `mapFileUseCount` and `mapDb`. After #9951 I'm going to refactor these to put them on a CDBWrapper object which represents a single database.
- If it is specific to wallets, the structure could be put on CWallet, and a reference passed in where it is needed (e.g. to CWalletDB) instead of looking it up all the time. This makes the CWallet more self-contained.

**Review by ryanofsky — COMMENTED** [review]
utACK 3f096ca2cad68c8bc66a15142dabce28ed8e6998

File: src/wallet/walletdb.cpp [hunk 7]
```diff
@@ -32,46 +32,38 @@ static std::atomic<unsigned int> nWalletDBUpdateCounter;
 
 bool CWalletDB::WriteName(const std::string& strAddress, const std::string& strName)
 {
-    nWalletDBUpdateCounter++;
```
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: 1. (thread 1) Increment the counter.
2. (thread 2) DB gets flushed
3. (thread 1) Actual change happens.
4. (thread 2) Thinks DB hasn't changed, so doesn't flush again.

File: src/wallet/walletdb.cpp [hunk 8]
```diff
@@ -103,42 +94,40 @@ bool CWalletDB::WriteCryptedKey(const CPubKey& vchPubKey,
         Erase(std::make_pair(std::string("key"), vchPubKey));
         Erase(std::make_pair(std::string("wkey"), vchPubKey));
     }
+
+    IncrementUpdateCounter();
```
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: Hmm, not sure what we want to happen in that case...

**Comment by luke-jr** [comment]
@laanwj I agree that stuff should be refactored, but I don't think multiwallet needs to depend on it here.

(Why is CWalletDB not just a private parent class of CWallet anyway?)

File: src/wallet/walletdb.cpp (hunk 8 above)
Comment (laanwj) [inline src/wallet/walletdb.cpp]: This is one of the reasons why I suggested above to only update the update counter if a DB transaction actually hits the database. That's more robust than doing it per operation. It's possible for all the operations to succeed then TxnAbort() to be called and still no update happens.

**Comment by laanwj** [comment]
> (Why is CWalletDB not just a private parent class of CWallet anyway?)

- It's named wrongly. Because CWalletDB is not a database, it provides access to it. To be exact, it is a database transaction, similar to the CDBBatch in leveldbwrapper.h. It is commited when the object goes out of scope or `TxnCommit()` is called. It is not committed if `TxnAbort()` is called.
- Should at least use composition there not inheritance. A wallet *uses* a wallet database, it *is* not a wallet database. This avoids low-level implementation details from leaking through.

> I agree that stuff should be refactored, but I don't think multiwallet needs to depend on it here.

As said I understand that. I'm not asking for you to do big changes. I just don't want it to become even worse and more muddled.

File: src/wallet/wallet.h [hunk 9]
```diff
@@ -30,7 +30,8 @@
 
 #include <boost/shared_ptr.hpp>
 
-extern CWallet* pwalletMain;
+typedef CWallet* CWallet_ptr;
```
Comment (TheBlueMatt) [inline src/wallet/wallet.h]: Why typedef this? Its literally /more/ characters now.

File: src/init.cpp [hunk 10]
```diff
@@ -256,8 +258,7 @@ void Shutdown()
 #endif
     UnregisterAllValidationInterfaces();
 #ifdef ENABLE_WALLET
-    delete pwalletMain;
-    pwalletMain = NULL;
+    vpwallets.clear();
```
Comment (TheBlueMatt) [inline src/init.cpp]: Don't you still need to delete each wallet here?

File: src/util.cpp (hunk 5 above)
Comment (TheBlueMatt) [inline src/util.cpp]: I believe this is also not correct - other threads may be using mapMultiArgs or the vector in it at the same time

**Review by TheBlueMatt — COMMENTED** [review]
Barely got started, will finish review later.

File: src/util.cpp (hunk 5 above)
Comment (TheBlueMatt) [inline src/util.cpp]: Maybe get @jtimon to do it in #9494 and then just use it here?

File: src/wallet/wallet.h (hunk 9 above)
Comment (laanwj) [inline src/wallet/wallet.h]: Typedef-ing a pointer type doesn't seem that useful to me either. I suppose the rationale is "what if we want to change it to e.g. std::shared_ptr later". But a) I think the code is more readable if the kind of pointer is clear from the function/struct/class signatures using it b) You might want to have different kinds of pointers to the same kind of objects in different places.

File: src/wallet/rpcwallet.cpp [hunk 11]
```diff
@@ -29,7 +29,8 @@
 
 CWallet *GetWalletForJSONRPCRequest(const JSONRPCRequest& request)
 {
-    return pwalletMain;
+    // TODO: Some way to access secondary wallets
```
Comment (NicolasDorier) [inline src/wallet/rpcwallet.cpp]: I throwing an RPC error like `RPC_METHOD_NOT_FOUND` might be better than returning `NULL`.

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -3724,24 +3734,17 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
 bool CWallet::InitLoadWallet()
 {
     if (GetBoolArg("-disablewallet", DEFAULT_DISABLE_WALLET)) {
-        pwalletMain = NULL;
```
Comment (NicolasDorier) [inline src/wallet/wallet.cpp]: `vpwallets.clear()` ?

File: src/wallet/wallet.h (hunk 9 above)
Comment (NicolasDorier) [inline src/wallet/wallet.h]: not sure it make sense either, but if it is done, better naming it to `CWalletRef` to be consistent with `CTransactionRef`

**Comment by NicolasDorier** [comment]
not blocking, but I think a `CWallets` instead of `vpwallets` it would make the code way easier to read and less error prone.
Would also make locking easier to get right, and preventing some loop.

It would impact:
```
std::map<std::string, CWalletDBStateInfo> WalletDBStateInfos
std::vector<CWallet_ptr> vpwallets;
IncrementUpdateCounter()
```
That said, this can be done in a separate PR.

File: src/util.cpp (hunk 5 above)
Comment (laanwj) [inline src/util.cpp]: > It can't get here unless the list is currently empty.

Good point

File: src/wallet/rpcwallet.cpp [hunk 13]
```diff
@@ -29,7 +29,8 @@
...
+    return vpwallets.empty() ? NULL : vpwallets[0];
```
Comment (laanwj) [inline src/wallet/rpcwallet.cpp]: s/NULL/nullptr?

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -442,30 +442,40 @@ bool CWallet::Verify()
...
+    SoftSetArg("-wallet", DEFAULT_WALLET_DAT);
+
     uiInterface.InitMessage(_("Verifying wallet..."));
-    std::string walletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
 
-    std::string strError;
-    if (!CWalletDB::VerifyEnvironment(walletFile, GetDataDir().string(), strError))
-        return InitError(strError);
+    for (const std::string& walletFile : mapMultiArgs.at("-wallet")) {
+        if (walletFile.find_first_of("/\\") != std::string::npos) {
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: `walletFile.find_first_of("/\\")` is not portable. Likely, boost::filesystem::path has a way to check a path for path separators.

File: src/wallet/walletdb.cpp (hunk 8 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: I agree, but it's outside the scope of this PR. For now, I'll just increment the counter more aggressively (merely reordering the increment to fix the race issue).

File: src/wallet/rpcwallet.cpp (hunk 11 above)
Comment (luke-jr) [inline src/wallet/rpcwallet.cpp]: It is not always an error for there to be no wallet (eg, signrawtransaction).

File: src/wallet/wallet.cpp (hunk 14 above)
Comment (luke-jr) [inline src/wallet/wallet.cpp]:
```C++
        boost::filesystem::path walletFileBoost = walletFile;
        if (walletFileBoost.has_root_path() || walletFileBoost.has_parent_path()) {
```

This would work, but off-topic here since this is existing code.

File: src/util.cpp (hunk 5 above)
Comment (luke-jr) [inline src/util.cpp]: @TheBlueMatt I don't see what your concern is... We have the lock here already.

**Comment by luke-jr** [comment]
In-scope nits addressed and rebased.

File: src/util.cpp [hunk 15]
```diff
@@ -484,6 +485,8 @@ void ForceSetArg(const std::string& strArg, const std::string& strValue)
 {
     LOCK(cs_args);
     mapArgs[strArg] = strValue;
+    _mapMultiArgs[strArg].clear();
```
Comment (jtimon) [inline src/util.cpp]: why clear here? afaik this is only used for unittests now.

Comment (luke-jr) [inline src/util.cpp]: So that there's only one value in _mapMultiArgs...

File: src/util.cpp (hunk 5 above)
Comment (TheBlueMatt) [inline src/util.cpp]: We do not take the lock when accessing mapMultiArgs all over the codebase. You cannot write a new entry to this map without updating the various locations in the codebase which access mapMultiArgs to also go through cs_args.

**Comment by luke-jr** [comment]
Rebased

File: src/wallet/walletdb.h [hunk 16]
```diff
@@ -247,6 +245,7 @@ class CWalletDB
     bool WriteVersion(int nVersion);
 private:
     CDB batch;
+    CWalletDBWrapper& wrapper;
```
Comment (ryanofsky) [inline src/wallet/walletdb.h]: In commit "CWalletDB: Store the update counter per wallet"

Might be nice to call this member dbw instead of wrapper for consistency with the wallet member.

File: src/wallet/walletdb.h [hunk 17]
```diff
@@ -140,6 +140,27 @@ class CKeyMetadata
...
 class CWalletDB
 {
+private:
+    template <typename K, typename T>
+    bool WriteIC(const K& key, const T& value, bool fOverwrite = true)
+    {
+        if (!batch.Write(key, value, fOverwrite)) {
+            return false;
+        }
+        IncrementUpdateCounter();
```
Comment (ryanofsky) [inline src/wallet/walletdb.h]: In commit "Bugfix: wallet: Increment "update counter" only after actually making the applicable db changes to avoid potential races"

Would be helpful to have a comment mentioning that counter should be updated after the write operation, rather than before to prevent a race where the flush thread could run and store the higher counter value before the database is fully updated.

Could also mention the idea of moving the counter increment to the destructor / commit function so these write and erase wrappers would be unnecessary.

File: src/wallet/db.h [hunk 18]
```diff
@@ -122,11 +122,15 @@ class CWalletDBWrapper
     void IncrementUpdateCounter();
     unsigned int GetUpdateCounter();
 
+    std::atomic<unsigned int> nUpdateCounter;
+    unsigned int nLastSeen;
+    unsigned int nLastFlushed;
+    int64_t nLastWalletUpdate;
```
Comment (ryanofsky) [inline src/wallet/db.h]: In commit "Wallet: Replace pwalletMain with a vector of wallet pointers"

It would be nice if conversion of these values from static variables to per-wallet members happened in the previous commit ("CWalletDB: Store the update counter per wallet") instead of this one. Would make both commits easier to follow and avoid the unusual intermediate state where globals and per-wallet members are compared to each other.

File: src/wallet/db.h [hunk 19]
```diff
@@ -122,11 +122,15 @@ class CWalletDBWrapper
     void IncrementUpdateCounter();
     unsigned int GetUpdateCounter();
 
+    std::atomic<unsigned int> nUpdateCounter;
+    unsigned int nLastSeen;
```
Comment (ryanofsky) [inline src/wallet/db.h]: In commit "Wallet: Replace pwalletMain with a vector of wallet pointers"

Should initialize these values.

File: src/wallet/db.h [hunk 20]
```diff
@@ -119,10 +119,14 @@ class CWalletDBWrapper
...
     void Flush(bool shutdown);
 
+    void IncrementUpdateCounter();
+    unsigned int GetUpdateCounter();
+
 private:
     /** BerkeleyDB specific */
     CDBEnv *env;
     std::string strFile;
+    std::atomic<unsigned int> nUpdateCounter;
```
Comment (ryanofsky) [inline src/wallet/db.h]: In commit "CWalletDB: Store the update counter per wallet":

Should initialize this value.

File: src/wallet/walletdb.cpp [hunk 21]
```diff
@@ -762,20 +760,22 @@ void MaybeCompactWalletDB()
         return;
     }
 
-    static unsigned int nLastSeen = CWalletDB::GetUpdateCounter();
-    static unsigned int nLastFlushed = CWalletDB::GetUpdateCounter();
+    CWalletDBWrapper& dbh = pwalletMain->GetDBHandle();
```
Comment (ryanofsky) [inline src/wallet/walletdb.cpp]: In commit "CWalletDB: Store the update counter per wallet"

Maybe call this dbw for consistency with existing wallet code.

**Review by ryanofsky — COMMENTED** [review]
utACK 68bc5b926fadbe1d904b5138dc299d2dd06fdead. This is much simpler than before and the bugfixes and cleanup are nice even independent of the multiwallet changes.

File: src/wallet/wallet.cpp [hunk 22]
```diff
@@ -440,30 +440,40 @@ bool CWallet::Verify()
...
+            return InitError(_("Invalid characters in -wallet filename"));
+        }
 
-    if (GetBoolArg("-salvagewallet", false))
-    {
-        // Recover readable keypairs:
-        CWallet dummyWallet;
-        if (!CWalletDB::Recover(walletFile, (void *)&dummyWallet, CWalletDB::RecoverKeysOnlyFilter))
+        std::string strError;
+        if (!CWalletDB::VerifyEnvironment(walletFile, GetDataDir().string(), strError))
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: This redundantly opens/checks the datadir via BDB. Not sure if this is a problem...

File: src/wallet/walletdb.cpp [hunk 23]
```diff
@@ -785,22 +760,23 @@ void MaybeCompactWalletDB()
         return;
     }
 
-    static unsigned int nLastSeen = CWalletDB::GetUpdateCounter();
-    static unsigned int nLastFlushed = CWalletDB::GetUpdateCounter();
-    static int64_t nLastWalletUpdate = GetTime();
+    for (CWalletRef pwallet : vpwallets) {
```
Comment (jonasschnelli) [inline src/wallet/walletdb.cpp]: Without a lock we set the assumption that `vpwallets` is immutable.

File: src/wallet/wallet.cpp [hunk 24]
```diff
@@ -440,30 +440,40 @@ bool CWallet::Verify()
     if (GetBoolArg("-disablewallet", DEFAULT_DISABLE_WALLET))
         return true;
 
+    SoftSetArg("-wallet", DEFAULT_WALLET_DAT);
+
     uiInterface.InitMessage(_("Verifying wallet..."));
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: using `Verifying wallet(s)...`?

File: src/wallet/wallet.cpp [hunk 25]
```diff
@@ -440,30 +440,40 @@ bool CWallet::Verify()
...
-        if (!CWalletDB::Recover(walletFile, (void *)&dummyWallet, CWalletDB::RecoverKeysOnlyFilter))
+        std::string strError;
+        if (!CWalletDB::VerifyEnvironment(walletFile, GetDataDir().string(), strError))
+            return InitError(strError);
+
+        if (GetBoolArg("-salvagewallet", false))
+        {
+            // Recover readable keypairs:
+            CWallet dummyWallet;
+            if (!CWalletDB::Recover(walletFile, (void *)&dummyWallet, CWalletDB::RecoverKeysOnlyFilter))
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: This renames an arbitrary wallet name to "wallet.%d.bak", collisions may happen.

`2017-05-24 07:56:04 Renamed test1.dat to wallet.1495612559.bak`

Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: Maybe we should use `<filename>.<date>.bak` (will result in something like `test1.dat.1495612559.bak` (which is preferable IMO).

**Comment by jonasschnelli** [comment]
Tested a bit and seems to work as intended.
The only point I found during testing that needs fixing is the SalvageWallet renaming to "wallet.%d.bak".
Step-debugged the flushing and SyncTransaction.

### Conceptual:
This PR mostly goes into a direction that one has to choose the wallets-to-work-with before starting up Core. I miss the pre-work for run-time adding and removing of wallets. Sure, this may be added later. 
Furthermore, the "globalness" of the startup arguments and action like `-usehd` or `-zapwallettxes`, `--keypool`, etc. are somehow conceptual issues that need to be solved before deploying multiwallet.

* Ideally creating, loading and unloading of wallets happens during runtime over an RPC call (or though the GUI).
* In the long run, actions like `zapwallettx`, `salvagewallet`, `rescan`, `upgradewallet`, etc should also be moved to runtime.
* Per wallet settings like `-usehd`, `-keypool`, etc. should be values that can be set during creation (or loading of a wallet).

**Comment by jonasschnelli** [comment]
Gitian builds are failing (all three platforms): https://bitcoin.jonasschnelli.ch/build/147

**Comment by gmaxwell** [comment]
> zapwallettx, salvagewallet

These should probably become less visible or go away or move to a separate utility. 

> rescan

We can do this from the rpc now.

Not sure about upgradewallet.

File: src/wallet/wallet.cpp (hunk 25 above)
Comment (jnewbery) [inline src/wallet/wallet.cpp]: I agree with @jonasschnelli . This seems dangerous if `Recover()` or `VerifyDatabaseFile()` is called for multiple wallets. Should be straightforward to update `Recover()` to name the backup file uniquely (the warning string in `VerifyDatabaseFile()` will also need to be updated)

**Review by jnewbery — COMMENTED** [review]
A couple of small nits inline, and a couple of general comments:

- I think this could do with both unit and integration tests
- Please update the comment for `CWallet::Verify()` in wallet.h now that it's responsible for parsing the wallet arguments and populating the ArgsManager

File: src/wallet/db.h (hunk 20 above)
Comment (luke-jr) [inline src/wallet/db.h]: It's an atomic, should have a default constructor?

Comment (ryanofsky) [inline src/wallet/db.h]: > It's an atomic, should have a default constructor?

Yes but it seems to not initialize the value, see
https://stackoverflow.com/questions/36320008/whats-the-default-value-for-a-stdatomic or http://en.cppreference.com/w/cpp/atomic/atomic/atomic. IMO, adding `= 0` to this line would make the code clearer anyway.

**Comment by luke-jr** [comment]
Rebased and issues addressed

**Comment by laanwj** [comment]
Travis falure:
```
Running 255 test cases...
unknown location(0): fatal error: in "wallet_tests/importwallet_rescan": unknown type
wallet/test/wallet_tests.cpp(446): last checkpoint: "importwallet_rescan" entry.
*** 1 failure is detected in the test module "Bitcoin Test Suite"
```

**Comment by ryanofsky** [comment]
> Travis falure:

The importwallet_rescan test has another pwalletMain usage that needs to be changed to vpwallets.

File: src/wallet/walletdb.cpp [hunk 26]
```diff
@@ -825,7 +825,7 @@ bool CWalletDB::VerifyEnvironment(const std::string& walletFile, const fs::path&
 
 bool CWalletDB::VerifyDatabaseFile(const std::string& walletFile, const fs::path& dataDir, std::string& warningStr, std::string& errorStr)
 {
-    return CDB::VerifyDatabaseFile(walletFile, dataDir, errorStr, warningStr, CWalletDB::Recover);
+    return CDB::VerifyDatabaseFile(walletFile, dataDir, warningStr, errorStr, CWalletDB::Recover);
```
Comment (ryanofsky) [inline src/wallet/walletdb.cpp]: In commit "Bugfix: wallet: Fix warningStr, errorStr argument order"

Note: the bug seems to have been introduced in #8574.

**Review by ryanofsky — COMMENTED** [review]
utACK 33f0b60b49305435c9a7fa353c91da15b3ca91a4 with the test fix. First 6 commits were basically unchanged since my previous review (just a member rename and atomic init fixes). 7 new commits have been added and seem fine.

**Comment by jnewbery** [comment]
I've had a quick scan and the new commits look generally good, although #10457 seems like a simpler solution to the wallet backup name issue. I don't think it's really worth passing the backup file name around through multiple function calls just so it can be printed in a single warning message.

I'll fully review once the the test has been fixed.

File: src/wallet/wallet.cpp [hunk 27]
```diff
@@ -3972,15 +3979,27 @@ bool CWallet::ParameterInteraction()
     }
 
     if (GetBoolArg("-salvagewallet", false) && SoftSetBoolArg("-rescan", true)) {
+        if (is_multiwallet) {
+            return InitError(strprintf("%s is only allowed with a single wallet file", "-salvagewallet"));
```
Comment (luke-jr) [inline src/wallet/wallet.cpp]: Both are correct. If we get to ::Verify, we should salvage all wallets. But for now, we forbid this combination. Moving it out of the loop is error-prone, and creates unnecessary complexity.

File: src/wallet/test/wallet_tests.cpp [hunk 28]
```diff
@@ -19,6 +19,8 @@
 #include <boost/test/unit_test.hpp>
 #include <univalue.h>
 
+extern CWallet* pwalletMain;
```
Comment (ryanofsky) [inline src/wallet/test/wallet_tests.cpp]: > This can probably be removed after the importwallet_rescan test is updated.

Never mind, it's used by the receiverequests test.

**Review by ryanofsky — COMMENTED** [review]
utACK c237bd750e11472f0f9639891036b975cf7faf15. Changes since previous review were fixing rescan test, updating a salvagewallet comment, and removing GetUpdateCounter in an earlier commit.

**Comment by laanwj** [comment]
utACK c237bd7
Seems ready to merge.

**Comment by laanwj** [comment]
> I've had a quick scan and the new commits look generally good, although #10457 seems like a simpler solution to the wallet backup name issue. I don't think it's really worth passing the backup file name around through multiple function calls just so it can be printed in a single warning message.

I agree. We should probably revert that part and do #10457 afterward.
